graphql_compiler/compiler/metadata.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2018-present Kensho Technologies, LLC.
+"""Utilities for recording, inspecting, and manipulating metadata collected during compilation."""
+from collections import namedtuple
+
+import six
+
+
+LocationInfo = namedtuple(
+    'LocationInfo',
+    (
+        'parent_location',         # Location/FoldScopeLocation, the parent of the current location
+        'type',                    # str, the actual type name at that location
+        'coerced_from_type',       # str, the type before coercion, or None if no coercion applied
+        'optional_scopes_depth',   # int, how many nested optional scopes this location is in
+        'recursive_scopes_depth',  # int, how many nested recursion scopes this location is in
+        'is_folded',               # bool, True if this location is within an optional scope
+    )
+)
+
+
+@six.python_2_unicode_compatible
+class QueryMetadataTable(object):
+    """Query metadata container with info on locations, inputs, outputs, and tags in the query."""
+
+    def __init__(self, root_location, root_location_info):
+        """Create a new empty QueryMetadataTable object."""
+        self._root_location = root_location  # Location, the root location of the entire query
+        self._locations = dict()             # dict, Location/FoldScopeLocation -> LocationInfo
+        self._inputs = dict()                # dict, input name -> input info namedtuple
+        self._outputs = dict()               # dict, output name -> output info namedtuple
+        self._tags = dict()                  # dict, tag name -> tag info namedtuple
+        self.register_location(root_location, root_location_info)
+
+    @property
+    def root_location(self):
+        """Return the root location of the query."""
+        return self._root_location
+
+    def register_location(self, location, location_info):
+        """Record a new location's metadata in the metadata table."""
+        old_info = self._locations.get(location, None)
+        if old_info is not None:
```

Comment:
Because `old_info` is used one more time in the error message. While a bit more convoluted, it ensures that the data used to decide to raise an error (`old_info`) is the exact same data that is reported in the error message.

Otherwise, the error message would have to do `self._locations[location]` one more time, which can introduce typos and assumes the `self._locations` dict hasn't changes in the meantime.